fairscale/experimental/nn/offload.py

```diff
@@ -432,27 +362,6 @@ def __init__(
         self._num_microbatches = num_microbatches
 
     def forward(self, *inputs: Any, **_: Any) -> Any:
-        dummy_input = torch.tensor([], requires_grad=True)
-        if self._checkpoint_activation:
-            return ActivationCheckpointing.apply(*inputs, dummy_input, self)
-
-        self._activations = []
-        for index in range(-1, len(self.model_slices)):
-            if index >= 0:
-                # TODO(anj-s): This might be a redundant call since we have the previous
-                # activation on the device already.
-                self._activations[index] = tuple([a.cuda() for a in list(self._activations[index])])
-                inputs = self._activations[index]
-                inputs = self.model_slices[index](*inputs)
-            # Call the custom autograd hooks (discard/load slices FW and BW)
-            inputs = ShardSyncLayer.apply(inputs, index, self.model_slices, self)
-            self._activations.append(inputs)
-            if index >= 0:
-                self._activations[index] = tuple([a.cpu() for a in list(self._activations[index])])
-
-        # We don't move the last activation/output since the target is present
-        # on the device.
-        # TODO(anj-s): It is now a requirement that the target tensors be placed on the
-        # device.
-        result = self._activations[-1]
-        return result[0] if len(result) == 1 else result
+        # We need the send param to be a dummy input to enable the
+        # backward pass to be triggered for integer inputs.
+        return ActivationCheckpointing.apply(*inputs, torch.tensor([], requires_grad=True), self)
```

Comment:
I am not totally understanding the magic here. Doesn't this end up calling self.forward() again, which will end up doing ActCkpt.apply() again?

Reply:
This calls self.forward of the ActivationCheckpointing class and correspondingly the self.backward of the same class.

Reply:
but checkpoint class's forward needs another forward function to work and the inputs, right? I thought the forward function as a parameter to the checkpoint class forward is "self", which is a callable. But that would end up calling this apply again, right?

Reply:
(flip a bool to mark in which scope you are ? sorry if obvious)

Reply:
Sorry I don't think I get the issue. I am implementing a custom autograd.Function with static forward and backward functions. I am not calling self() but instead self.layer_shard() so it should trigger the forward pass of the shard/module object.

Reply:
I don't know there is an "issue" or not. The only question here is asking what's this code doing? It is not clear to me what is it doing, that's all. Maybe very detailed comments?

Reply:
Added comments though the details are present in the documentation for autograd.Function which I have linked to. I don't think it makes sense to explain the details about the [apply hook](https://pytorch.org/docs/stable/_modules/torch/autograd/function.html#Function).